### PR TITLE
msgchan:add message channels

### PR DESCRIPTION
**Pull request by KeithWiles**
Adding support for message channels using two lockless rings. The rings
are used to send and receive data from a thread to another thread. The go
binding code needs a method to communicate between Go <--> C in a fast
and simple way. The library is not limited to Go <--> C and can be used
to create a pipe like interface between two or more threads.

The msgchan code mc_create() allows for one thread to create the channel, which
allows other threads to locate and attach to the msgchan as children.

Signed-off-by: Keith Wiles <keith.wiles@intel.com>

### CODE REVIEW

File: lib/core/msgchan/msgchan_priv.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    int child_count;                 /**< Number of children */
+    TAILQ_HEAD(, msg_chan) children; /**< List of attached children */
+    uint64_t send_calls;             /**< Number of send calls */
+    uint64_t send_cnt;               /**< Number of objects sent */
+    uint64_t recv_calls;             /**< Number of receive calls */
+    uint64_t recv_cnt;               /**< Number of objects received */
+    uint64_t recv_timeouts;          /**< Number of receive timeouts */
+} msg_chan_t;
+
+#define MC_CREATE_RECURSIVE_MUTEX(mutex)                                      \
```
Comment (jeffreybshaw) [inline lib/core/msgchan/msgchan_priv.h:46]: Can this be a functino? Doesn't seem like it needs to be a macro to me. Also, the caller might wnat to know if any of these functinos fail as it should'nt continue processing.

File: lib/core/meson.build [hunk 2]
```diff
@@ -26,6 +26,7 @@ dirs = [
 	'pktdev',
 	'txbuff',
     'pmds',
+    'msgchan',
```
Comment (jeffreybshaw) [inline lib/core/meson.build:29]: This seems more like an application library than a core library to me.

**Review by jeffreybshaw — CHANGES_REQUESTED** [review]
Can you update documentation too?

I didn't review the actual code yet, just a couple of quick comments.

File: lib/core/msgchan/msgchan_priv.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    do {                                                                      \
+        pthread_mutexattr_t attr;                                             \
+        if (pthread_mutexattr_init(&attr))                                    \
+            cne_print("%s: Unable to init mutex attributes\n", __func__);     \
+        else if (pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE))   \
+            cne_printf("%s: Unable to set mutex attributes\n", __func__);     \
+        else if (pthread_mutex_init(&(mutex), &attr)) {                       \
+            (void)pthread_mutexattr_destroy(&attr);                           \
```
[the remark this answers is not in the document]
Comment (KeithWiles) [reply lib/core/msgchan/msgchan_priv.h]: Fixed

File: lib/core/msgchan/msgchan_priv.h (hunk 1 above)
Comment (KeithWiles) [reply lib/core/msgchan/msgchan_priv.h]: Create a function

**Comment by KeithWiles** [comment]
I did a force commit as the files moved and thought it would be easier to review.

**Comment by KeithWiles** [comment]
We need to squash this one when ready

File: lib/usr/clib/msgchan/msgchan.h [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+ *
+ * @param mc
+ *   The message channel structure pointer
+ * @return
+ *   NULL if invalid pointer or string to message channel name
+ */
+CNDP_API const char *mc_name(msgchan_t *mc);
+
+/**
+ * @brief Return size and free space in the server/Client rings.
```
Comment (maryamtahhan) [inline lib/usr/clib/msgchan/msgchan.h:114]: not sure if we want to stick to the producer/consumer nomenclature vs the server/client... what do you think?

File: lib/usr/clib/msgchan/msgchan_priv.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+#define _MSGCHAN_PRIV_H_
+
+#include <sys/queue.h>
+#include <cne_common.h>
+#include <cne_ring.h>
+#include <cne_ring_api.h>
+
+/**
+ * @file
+ * Message Channels
```
[the remark this answers is not in the document]
Comment (KeithWiles) [reply lib/usr/clib/msgchan/msgchan_priv.h:15]: I changed the text. I ran doxygen and did not see a problem.

File: lib/usr/clib/msgchan/msgchan.c [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+    mc = calloc(1, sizeof(msg_chan_t));
+    if (!mc)
+        CNE_ERR_GOTO(err, "Unable to allocate memory\n");
+
+    mc->cookie = MC_COOKIE;
+
+    n = strlcpy(mc->name, "P:", sizeof(mc->name));
+    strlcpy(mc->name + n, name, sizeof(mc->name) - n);
+
+    n = strlcpy(rname, "SR:", sizeof(rname));
```
[the remark this answers is not in the document]
Comment (KeithWiles) [reply lib/usr/clib/msgchan/msgchan.c:164]: Yes, they are backwards.

File: lib/usr/clib/msgchan/msgchan.c [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+    }
+}
+
+static int
+__recv(msg_chan_t *mc, void **objs, int count, uint64_t msec)
+{
+    cne_ring_t *r;
+    int nb_objs = 0;
+
+    mc->recv_calls++;
```
[the remark this answers is not in the document]
Comment (maryamtahhan) [reply lib/usr/clib/msgchan/msgchan.c:250]: > Would it be a problem if msgchan had a restriction or only one thread per parent or child msgchan?

I think it's fine. If in the future there's a requirement to change that the stats can be modified to an atomic as you mention

**Review by maryamtahhan — APPROVED** [review]
LGTM

**Comment by KeithWiles** [comment]
I added some documentation and updated the requested changes.

**Pull request merged**